### ISSUE TITLE
🚀 Feature: More flexible dev server based proxy instead static config in `package.json`

### ISSUE THREAD

**Issue by JounQin** [backstage/backstage#25960]
### 🔖 Feature description

As I read the source codes at https://github.com/backstage/backstage/blob/57cb03d6b6e7ff8998c377c3d1bb94e1c49ec043/packages/cli/src/lib/bundler/server.ts#L217

It seems the proxy config can only be defined at `app`'s `package.json` file which must be static, that limits this feature very much, can we read `app-config*` files for this config at the same time?

### 🎤 Context

Sometimes we want to change the proxy target for different target for development, but that doesn't need to be committed every time, if we can read dev server proxy from `app-config*` also, that would be better for such cases.

### ✌️ Possible Implementation

Just add maybe `app.devServer.proxy` config?

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by Rugvip** [comment]
+1 to more powerful proxying features for local development in general, although not exactly sure where and what that should look like at this point. Another option is to have the backend itself do a bit more heavy-lifting here, rather than building things into the dev server.

**Comment by JounQin** [comment]
Yes, I'm using backend proxy right now, but that means we'll have those backend proxies in production also which could be unnecessary actually.

**Comment by JounQin** [comment]
ping

**Comment by Rugvip** [comment]
Might look at this in a lil bit but pretty busy with other prios at the moment. In the meantime, you could split out some config into for example `app-config.development.yaml` or clear the proxy endpoints in `app-config.production.yaml` with `null` values in order to avoid having these endpoints available in production.

**Comment by JounQin** [comment]
@Rugvip Can we reopen this issue?

**Comment by JounQin** [comment]
bump

**Comment by JounQin** [comment]
nope

**Comment by camilaibs** [comment]
Hi @JounQin 👋🏻 , would you have a moment to try implementing this and create a pull request? A contribution would be greatly appreciated.
